modules/layers/src/solid-polygon-layer/polygon.js

```diff
@@ -47,86 +52,217 @@ function isClosed(simplePolygon) {
+    target[targetIndex++] = positions[srcStartIndex + i];
+  }
+
+  if (!isFlatPathClosed(positions, size, srcStartIndex, srcEndIndex)) {
+    for (let i = 0; i < size; i++) {
+      target[targetIndex++] = positions[srcStartIndex + i];
+    }
+  }
+  return targetIndex;
 }
 
 /**
- * Check if this is a non-nested polygon (i.e. the first element of the first element is a number)
- * @param {Array} polygon - either a complex or simple polygon
- * @return {Boolean} - true if the polygon is a simple polygon (i.e. not an array of polygons)
+ * Counts the number of vertices in a simple polygon, closes the polygon if needed.
+ * @param {Array} simplePolygon - array of points
+ * @returns {Number} vertex count
  */
-export function getVertexCount(polygon) {
-  return isSimple(polygon)
-    ? _getVertexCount(polygon)
-    : polygon.reduce((length, simplePolygon) => length + _getVertexCount(simplePolygon), 0);
+function getNestedVertexCount(simplePolygon) {
+  return (isNestedPathClosed(simplePolygon) ? 0 : 1) + simplePolygon.length;
 }
 
-// Roughly estimate the number of triangles needed to tesselate the polygon
-// For allocating the index buffer - the number returned might be excessive
-export function getTriangleCount(complexPolygon) {
-  let triangleCount = 0;
-  let first = true;
-  for (const simplePolygon of complexPolygon) {
-    const size = simplePolygon.length;
-    if (first) {
-      triangleCount += size >= 3 ? size - 2 : 0;
-    } else {
-      triangleCount += size + 1;
+/**
+ * Counts the number of vertices in a simple flat array, closes the polygon if needed.
+ * @param {Array} positions - array of numbers
+ * @param {Number} size - size of a position, 2 (xy) or 3 (xyz)
+ * @param {Number} [startIndex] - start index of the path in the positions array
+ * @param {Number} [endIndex] - end index of the path in the positions array
+ * @returns {Number} vertex count
+ */
+function getFlatVertexCount(positions, size, startIndex = 0, endIndex) {
+  endIndex = endIndex || positions.length;
+  if (startIndex >= endIndex) {
+    return 0;
+  }
+  return (
+    (isFlatPathClosed(positions, size, startIndex, endIndex) ? 0 : 1) +
+    (endIndex - startIndex) / size
+  );
+}
+
+/**
+ * Counts the number of vertices in any polygon representation.
+ * @param {Array|Object} polygon
+ * @param {Number} positionSize - size of a position, 2 (xy) or 3 (xyz)
+ * @returns {Number} vertex count
+ */
+export function getVertexCount(polygon, positionSize) {
+  if (polygon.positions) {
+    // complex flat
+    const {positions, loopStartIndices} = polygon;
+
+    if (loopStartIndices) {
+      let vertexCount = 0;
+      for (let i = 0; i < loopStartIndices.length; i++) {
+        vertexCount += getFlatVertexCount(
+          polygon.positions,
+          positionSize,
+          loopStartIndices[i],
+          loopStartIndices[i + 1]
+        );
+      }
+      return vertexCount;
     }
-    first = false;
+    polygon = positions;
```

Comment:
Is it safe to mutate the input?

Reply:
Not mutating the input, only changing the variable in the function's scope.